explorer/interpreter/impl_scope.cpp

```diff
@@ -415,10 +415,10 @@ auto ImplScope::TryResolveInterfaceHere(
 
 // TODO: Add indentation when printing the parents.
 void ImplScope::Print(llvm::raw_ostream& out) const {
-  out << "impl declarations: ";
   llvm::ListSeparator sep;
```

Comment:
I wonder if it would be better to use a different separator here. The current lists of impl declarations are very hard to read and lead to extremely long lines. Maybe something like this:
```suggestion
  llvm::ListSeparator sep(",\n    ");
```

I'm thinking about output something like this:
```
--- impl declarations for `interface TestInterface` (full_trace.carbon:7)
 + []
 + [`bool` as `interface EqWith(U = bool)`,
    `i32` as `interface EqWith(U = i32)`,
    `String` as `interface EqWith(U = String)`,
    `i32` as `interface CompareWith(U = i32)`,
    `String` as `interface CompareWith(U = String)`,
    `i32` as `interface LessWith(U = i32)`,
    `String` as `interface LessWith(U = String)`,
    `i32` as `interface LessEqWith(U = i32)`,
    `String` as `interface LessEqWith(U = String)`,
    `i32` as `interface GreaterWith(U = i32)`,
    `String` as `interface GreaterWith(U = String)`,
    `i32` as `interface GreaterEqWith(U = i32)`,
    `String` as `interface GreaterEqWith(U = String)`,
    `i32` as `interface Negate`,
    `i32` as `interface AddWith(U = i32)`,
    `i32` as `interface SubWith(U = i32)`,
    `i32` as `interface MulWith(U = i32)`,
    `i32` as `interface DivWith(U = i32)`,
    `i32` as `interface ModWith(U = i32)`,
    `i32` as `interface BitComplement`,
    `i32` as `interface BitAndWith(U = i32)`,
    `i32` as `interface BitOrWith(U = i32)`,
    `i32` as `interface BitXorWith(U = i32)`,
    `i32` as `interface LeftShiftWith(U = i32)`,
    `i32` as `interface RightShiftWith(U = i32)`,
    `i32` as `interface Inc`,
    `i32` as `interface Dec`,
    `U` as `interface __EqualConverter` [0],
    `(T1,)` as `interface As(T = (U1,))` [0, 0, 0; 1, 1, 0, 0, 0],
    `(T1, T2)` as `interface As(T = (U1, U2))` [0, 0, 0; 0, 0, 1; 1, 1, 0, 0, 0; 1, 1, 0, 0, 1],
    `(T1, U1)` as `interface EqWith(U = (T2, U2))` [0, 0, 0; 0, 0, 1; 1, 1, 0, 0, 0; 1, 1, 0, 0, 1],
    `(T1, T2, T3)` as `interface As(T = (U1, U2, U3))` [0, 0, 0; 0, 0, 1; 0, 0, 2; 1, 1, 0, 0, 0; 1, 1, 0, 0, 1; 1, 1, 0, 0, 2],
    `T` as `interface ImplicitAs(T = U)` [0; 1, 1, 0],
    `T` as `interface As(T = U)` [0; 1, 1, 0],
    `T` as `interface ImplicitAs(T = U)` [0; 1, 1, 0],
    `T` as `interface As(T = U)` [0; 1, 1, 0],
    `T` as `interface AssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface AddAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface SubAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface MulAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface DivAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface ModAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface BitAndAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface BitOrAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface BitXorAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface LeftShiftAssignWith(U = U)` [0; 1, 1, 0],
    `T` as `interface RightShiftAssignWith(U = U)` [0; 1, 1, 0]]
```

... which I think is a lot more readable. (Though it's very long! I think we should be printing this out a lot less than we currently do. But we can leave any improvements there to a different PR.)

Reply:
I think it is much better this way.
I was thinking of adding source location information to `ImplFact` to deal with this issue and filter is based on file context but I'm not able to figure it out yet.